src/Tasks/Microsoft.Common.CurrentVersion.targets

```diff
@@ -720,6 +720,120 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     </EmbeddedResource>
   </ItemDefinitionGroup>
 
+  <!--
+    ========================
+    DeterministicSourcePaths
+    ========================
+    
+    Unless specified otherwise enable deterministic source paths when building deterministically on CI server, but not for local builds.
+    In order for the debugger to find source files when debugging a locally built binary the PDB must contain original, unmapped local paths.
+  -->
+  <PropertyGroup>
+    <DeterministicSourcePaths Condition="'$(DeterministicSourcePaths)' == '' and '$(Deterministic)' == 'true' and '$(ContinuousIntegrationBuild)' == 'true'">true</DeterministicSourcePaths>
+  </PropertyGroup>
+
+  <!--
+    ==========
+    SourceRoot
+    ==========
+    
+    All source files of the project are expected to be located under one of the directories specified by SourceRoot item group.
+    This targets collects all SourceRoots from various sources (InitializeSourceRootDependsOn targets).
+    
+    If DeterministicSourcePaths is true it calculates deterministic mapped paths that can be used by CoreCompile targets
+    to initialize deterministic path mapping specific to the compiler.
+    
+    Include the final backslash in the path of each SourceRoot item.
+    
+    Extensibility: Add target to InitializeSourceRootDependsOn that populates SourceRoot item group.
+  -->
+
+  <Target Name="InitializeSourceRoot"
+          DependsOnTargets="$(InitializeSourceRootDependsOn)">
+    <ItemGroup>
+      <_InvalidSourceRoot Include="@(SourceRoot)" Condition="!HasTrailingSlash('%(SourceRoot.Identity)')" />
+    </ItemGroup>
+
+    <!--
+      The SourceRoot is required to have a trailing directory separator.
+      We do not append one implicitly as we do not know which separator to append on Windows.
+      The usage of SourceRoot might be sensitive to what kind of separator is used (e.g. in SourceLink where it needs
+      to match the corresponding separators used in paths given to the compiler).
+    -->
+    <Error Text="SourceRoot paths are required to end with a slash or backslash: @(_InvalidSourceRoot, ', ')"
+           Condition="'@(_InvalidSourceRoot)' != ''" />
+
+    <!--
+      Assign top-level source roots a deterministic path (MappedPath metadata).
+    -->
+
+    <ItemGroup Condition="'@(_MappedSourceRoot)' != ''">
+      <_MappedSourceRoot Remove="@(_MappedSourceRoot)" />
```

Comment:
Is this just defensive? That's not something we generally do.

Reply:
Yes, it's defensive since `<Output>` with `ItemName` adds to the list. Shouldn't we do it more generally? What if someone has a temporary item group of the same name?